### PR TITLE
fix: no data issue 18

### PR DESCRIPTION
**Pull request by swelf19**
PR solves the lack of data problem that was caused by Prometheus timeout when requesting metrics.
all monitors processed requests by retrieving data on demand. At the moment, two monitors: Slashing monitor and UpdateGlobalIndexBotMonitor take a long time to process the request.
Now these two monitors receive data from the terra lcd servers in the background and give it back to Prometheus without wasting time on requests in real time.

### CODE REVIEW

File: collector/monitors/slashing_monitor.go [hunk 1]
```diff
@@ -28,41 +29,74 @@ const (
...
+		m.flowManager = make(chan struct{})
+	}
+
+	go m.Do()
+
 	return m
 }
 
+func (m *SlashingMonitor) Do() {
+	for range m.flowManager {
```
Comment (oopcode) [inline collector/monitors/slashing_monitor.go:63]: I think this is a little bit overengineered. The standard way looks like this:

```
tk := time.NewTicker(time.Second) // From configuration.
	for {
		select {
		case <-tk.C:
			// Move the mutex inside updateData.
			err := m.updateData(context.Background())
			if err != nil {
				m.logger.Errorln("failed to update SlashingMonitor data:", err)
			}
		case <-ctx.Done():
			return
		}
	}
```

File: collector/monitors/slashing_monitor.go [hunk 2]
```diff
@@ -28,41 +29,74 @@ const (
...
-		metrics:              make(map[MetricName]float64),
+		metrics:              make(map[MetricName]MetricValue),
 		metricVectors:        make(map[MetricName]MetricVector),
 		apiClient:            cfg.GetTerraClient(),
 		validatorsRepository: repository,
 		logger:               cfg.Logger,
+		lock:                 sync.RWMutex{},
 	}
 
+	if cfg.SlashingMonitorUpdateInterval > 0 {
```
Comment (oopcode) [inline collector/monitors/slashing_monitor.go:51]: I think this is a bit overengineered. This is sort of the standard way to do such things:

```
func NewSlashingMonitor(cfg config.CollectorConfig, repository ValidatorsRepository) *SlashingMonitor {
	m := &SlashingMonitor{
		metrics:              make(map[MetricName]MetricValue),
		metricVectors:        make(map[MetricName]MetricVector),
		apiClient:            cfg.GetTerraClient(),
		validatorsRepository: repository,
		logger:               cfg.Logger,
		lock:                 sync.RWMutex{},
	}

	if cfg.SlashingMonitorUpdateInterval > 0 {
		go m.updateRoutine(ctx, time.Second*time.Duration(cfg.SlashingMonitorUpdateInterval))
	}

	return m
}

func (m *SlashingMonitor) updateRoutine(ctx context.Context, updateInterval time.Duration) {
	tk := time.NewTicker(updateInterval)
	for {
		select {
		case <-tk.C:
			// Move the mutex inside updateData.
			err := m.updateData(context.Background())
			if err != nil {
				m.logger.Errorln("failed to update SlashingMonitor data:", err)
			}
		case <-ctx.Done():
			return
		}
	}
}
```

File: collector/config/flow_namager.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package config
+
+import "time"
+
+func FlowManager(interval uint) chan struct{} {
```
Comment (oopcode) [inline collector/config/flow_namager.go:5]: I don't really see the need for this type.

**Comment by oopcode** [comment]
Are we sure that we need different update mechanics for different monitors? Maybe all monitors should be updated in background?

**Review by oopcode — CHANGES_REQUESTED** [review]
Please simplify the background routines and consider having 1 updates logic for all monitors.

**Review by pr0n00gler — CHANGES_REQUESTED** [review]
Also there are some warnings from `go vet` about locks:
```
MacBook-Pro-Mihail-2:terra-monitors pr0n00gler$ go vet ./...
# github.com/lidofinance/terra-monitors/collector/monitors
collector/monitors/bluna_token_info_monitor.go:31:9: return copies lock value: github.com/lidofinance/terra-monitors/collector/monitors.BlunaTokenInfoMonitor contains sync.RWMutex
collector/monitors/bluna_token_info_monitor.go:43:9: Name passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.BlunaTokenInfoMonitor contains sync.RWMutex
collector/monitors/bluna_token_info_monitor.go:97:9: GetMetrics passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.BlunaTokenInfoMonitor contains sync.RWMutex
collector/monitors/bluna_token_info_monitor.go:103:9: GetMetricVectors passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.BlunaTokenInfoMonitor contains sync.RWMutex
collector/monitors/monitors.go:45:10: Labels passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.MetricVector contains sync.RWMutex
collector/monitors/slashing_monitor.go:157:9: GetMetricVectors passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.SlashingMonitor contains sync.RWMutex
collector/monitors/update_global_index_monitor.go:59:9: Name passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.UpdateGlobalIndexMonitor contains sync.RWMutex
collector/monitors/update_global_index_monitor.go:160:9: GetMetrics passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.UpdateGlobalIndexMonitor contains sync.RWMutex
collector/monitors/update_global_index_monitor.go:166:9: GetMetricVectors passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.UpdateGlobalIndexMonitor contains sync.RWMutex
collector/monitors/validators_commission_monitor.go:40:9: return copies lock value: github.com/lidofinance/terra-monitors/collector/monitors.ValidatorsCommissionMonitor contains sync.Mutex
collector/monitors/validators_commission_monitor.go:81:9: GetMetricVectors passes lock by value: github.com/lidofinance/terra-monitors/collector/monitors.ValidatorsCommissionMonitor contains sync.Mutex
```

File: collector/monitors/validators_commission_monitor.go [hunk 4]
```diff
@@ -56,16 +63,21 @@ func (m *ValidatorsCommissionMonitor) Handler(ctx context.Context) error {
...
 	return nil
 }
 
-func (m *ValidatorsCommissionMonitor) GetMetrics() map[MetricName]float64 {
+func (m *ValidatorsCommissionMonitor) GetMetrics() map[MetricName]MetricValue {
 	return m.metrics
 }
 
-func (m ValidatorsCommissionMonitor) GetMetricVectors() map[MetricName]MetricVector {
+func (m ValidatorsCommissionMonitor) GetMetricVectors() map[MetricName]*MetricVector {
```
[the remark this answers is not in the document]
Comment (swelf19) [reply collector/monitors/validators_commission_monitor.go]: done

**Comment by swelf19** [comment]
> Also there are some warnings from `go vet` about locks:
> 
fixed

**Review by oopcode — APPROVED** [review]

File: collector/config/config.go [hunk 5]
```diff
@@ -16,19 +16,23 @@ const (
 	DefaultValidatorRegistryAddress = "terra1aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
 	DefaultAirDropRegistryContract  = "terra_dummy_airdrop"
 	DefaultRewardDispatcherContract = "terra_dummy_rewarddispatcher"
+
+	DefaultUpdateDataInterval = 30
```
Comment (oopcode) [inline collector/config/config.go:20]: You can set this as a duration right away, btw

File: collector/monitors/slashing_monitor.go [hunk 6]
```diff
@@ -25,42 +26,58 @@ const (
 )
 
 type SlashingMonitor struct {
-	metrics              map[MetricName]float64
-	metricVectors        map[MetricName]MetricVector
+	metrics       map[MetricName]MetricValue
+	metricVectors map[MetricName]*MetricVector
+	// tmp* for 2stage nonblocking update data
```
Comment (oopcode) [inline collector/monitors/slashing_monitor.go:31]: I don't think we need these values at the struct level? They can easily be declared as local variables inside `Handler()`.

**Review by oopcode — CHANGES_REQUESTED** [review]
pls get rid of the redundant tmp struct fields

File: collector/config/config.go (hunk 5 above)
Comment (swelf19) [reply collector/config/config.go]: done

**Review by oopcode — APPROVED** [review]

**Pull request merged**